src/metabase/api/collection.clj

```diff
@@ -45,17 +45,54 @@
 
 (declare root-collection)
 
+(defn- remove-other-users-personal-collections
+  "Remove from the provided collections any that are personal and belong to another user.
+   Will work in a single pass if collections are already ordered such that root (location = /)
+   collections appear before children, but will catch any mis-ordered entries and
+   do a final pass on those if needed."
+  ([user-id collections]
+   (loop [[{:keys [id location personal_owner_id] :as c} & r] collections
+          keep-roots     #{user-id}
+          skip-roots     #{}
+          res            []
+          indeterminates []]
+     (if c
+       (if (= "/" location)
+         (if (or (nil? personal_owner_id) (= user-id personal_owner_id))
+           (recur r (conj keep-roots id) skip-roots (conj res c) indeterminates)
+           (recur r keep-roots (conj skip-roots id) res indeterminates))
+         (let [[_ root] (str/split location #"/")
```

Comment:
we have a helper fn to explode the location in `metabase.models.collection/location-path->ids` maybe it's better to use that fn so we don't have to do the same thing of split->parse over and over.

Reply:
Done